### PR TITLE
Use Data.Compact to serialize K definition at compilation time

### LINKED ISSUE #2943: Separate out the definition loading and definition verifying routines
Right now, when I run speedscope on the invocation of `kore-match-disjunction` generated by running `krun tests/free-variable-capture.lambda --search --pattern '<k> V:K </k>'` on the k frontend definition `k-distribution/pl-tutorial/1_k/3_lambda++/lesson_2`, it reports that 70% of execution time is spent in `loadDefinitions`. Of that time, 40% of the overall total is spent loading and verifying the kore definition. This amounts to roughly 360 milliseconds since when run without profiling enabled, the application takes 900ms total.

Unfortunately, much of this time is likely being spent doing validation and verification of the kore definition (I assume; it seems to me that indexing the definition ought to be very fast and take less than 100ms on a definition of this size, but I don't fully understand the data structures the haskell backend uses internally). Assuming my assumption is correct, we ought to be able to do away with 100% of that time. This is because kompile already invokes the Haskell backend to validate and verify the kore definition, so it doesn't need to happen again each time that we invoke `kore-exec` or `kore-match-disjunction`.

However, currently these two functions (verifying and loading) are very tightly intercoupled. We ought to instead have separate routines for loading and verifying definitions. The verification ought to only happen at kompile time, whereas the loading ought to be extremely quick.

### PR DESCRIPTION
**Pull request by dwightguth**
Fixes #2943 

### Scope:

This change makes use of the Data.Compact library to serialize the initialized KORE definition ahead of time and then load it into memory at the start of kore-exec for symbolic execution and search. Nothing necessarily prevents us from using it for other execution modes as well, but I figured we would start with the basic case and we can add on more cases in follow-up changes if we so desire.

Note: the integration tests do not currently test this code path! It is only exercised if krun passes the path to a binary serialized file instead of the path to definition.kore. this requires frontend change that cannot be merged until this change is in master. Thus, the integration tests passing with this change indicates that the old code path still works. I have manually tested the new code path with the modified frontend and it did pass the integration tests.

**DO NOT MERGE** unless the head commit of this branch as of merge time passes the integration tests manually with the modified k frontend in addition to all checks being green.

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/app/exec/Main.hs [hunk 1]
```diff
@@ -620,59 +635,87 @@ mainDispatch = warnProductivity . mainDispatchWorker
                 then koreBmc localOptions proveOptions
                 else koreProve localOptions proveOptions
         | Just searchOptions <- koreSearchOptions =
-            koreSearch localOptions searchOptions
+            koreSearch exeLastModifiedTime localOptions searchOptions
+        | True <- serialize =
+            koreSerialize localOptions
```
Comment (ana-pantilie) [inline kore/app/exec/Main.hs:640]: So if you give it the serialize flag it will only serialize the definition.

File: kore/app/exec/Main.hs [hunk 2]
```diff
@@ -620,59 +635,87 @@ mainDispatch = warnProductivity . mainDispatchWorker
...
     Main (KFileLocations, ExitCode)
-koreSearch LocalOptions{execOptions, simplifierx} searchOptions = do
+koreSearch exeLastModifiedTime LocalOptions{execOptions, simplifierx} searchOptions = do
     let KoreExecOptions{definitionFileName} = execOptions
-    definition <- loadDefinitions [definitionFileName]
     let KoreExecOptions{mainModuleName} = execOptions
-    mainModule <- loadModule mainModuleName definition
+    let KoreExecOptions{koreSolverOptions} = execOptions
+    SerializedDefinition{serializedModule, lemmas, locations} <-
+        deserializeDefinition simplifierx koreSolverOptions definitionFileName mainModuleName exeLastModifiedTime
```
Comment (ana-pantilie) [inline kore/app/exec/Main.hs:658]: If the definition wasn't serialized before, it will serialize it by default?

**Review by ana-pantilie — COMMENTED** [review]
Can you explain how this will integrate with the frontend? I just want an example use case scenario, just so I understand how all parts fit together, this is what I've been having most trouble with when reviewing.

File: kore/app/share/GlobalMain.hs [hunk 3]
```diff
@@ -487,7 +541,112 @@ mainParse parser fileName = do
...
+    solverOptions
+    definitionFilePath
+    mainModuleName
+    exeLastModifiedTime =
+        do
+            bytes <- ByteString.readFile definitionFilePath & liftIO
+            let magicBytes = ByteString.drop 8 $ ByteString.take 16 bytes
+            let magic = Binary.decode @Word64 magicBytes
+            case magic of
+                0x7c155e7a53f094f2 -> do
```
[the remark this answers is not in the document]
Comment (dwightguth) [reply kore/app/share/GlobalMain.hs:610]: This is the magic number that the Data.Compact package emits at the top of files that it serializes to in order to do a basic integrity check. They do not export that symbol so we had to recreate it here. This is used to detect whether the input is a serialized definition or a textual kore definition.

File: kore/app/exec/Main.hs (hunk 1 above)
Comment (dwightguth) [reply kore/app/exec/Main.hs:640]: Well, a lot of the logic needed to do this was part of the Kore.Exec and kore/app/exec/Main files, so it seemed more appropriate to put it here than to do a huge refactoring of where a whole bunch of code lived. As it was, I already had to move around a fair bit.

File: kore/app/exec/Main.hs (hunk 2 above)
Comment (dwightguth) [reply kore/app/exec/Main.hs:658]: I'm not entirely sure what you're asking. the `deserializeDefinition` function returns a `SerializedDefinition` object which contains everything needed to do execution or search with the haskell backend. What it does is detect whether the file it's reading is a serialized binary file or a textual kore definition and if it is the former, it simply deserializes it, and if it is the latter, it does all the work done by `kore-exec --serialize`, but does not actually write anything to disk, merely using the result to do execution.

**Comment by dwightguth** [comment]
Here is the changes that will be made to the frontend after this is merged: https://github.com/runtimeverification/k/pull/2551

As you can see in that diff, all the frontend is really doing differently is calling `kore-exec --serialize` instead of `kore-parser`. Since `kore-exec --serialize` already parses and verifies the definition, all we're doing is adding the extra step of compacting and serializing it when that's done.

As far as performance, this takes the time to execute a single step of the haskell backend on the c semantics down from like 40 seconds to 7 seconds. That's a pretty substantial improvement. It also sizeably increases the GC throughput in the backend while executing because it puts the initialized KORE definition in what's called a compact region, which essentially means, among other things, that it's not being traced by the garbage collector.

You are correct that it probably does increase the memory usage somewhat, because one consequence of being a compact region is that any pointer into any part of a compact region keeps the entire region alive and prevents garbage collection. But overall as you can see, the difference is relatively minor, and the actual execution time improves substantially. compilation does take longer, but in doing so it saves substantial amounts of time on the fast path, so overall you benefit greatly unless you are compiling a lot of definitions and then never using them. So overall it is a substantial win.

**Review by ana-pantilie — APPROVED** [review]
Looks good, just one small request: where I asked specific questions about added/modified code, please add documentation for that as comments in the code. Thank you.

**Pull request merged**